packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -0,0 +1,182 @@
+      lovelaceSupply: () => firstValueFrom(this.networkInfo.lovelaceSupply),
+      protocolParameters: () => firstValueFrom(this.networkInfo.protocolParameters),
+      stake: () => firstValueFrom(this.networkInfo.stake)
+    };
+
+    this.connect();
+  }
+
+  private connect() {
+    this.status = 'connecting';
+
+    const ws = new WebSocket(this.url);
+
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    ws.onmessage = (event) => {
+      try {
+        if (typeof event.data !== 'string') throw new Error('Unexpected data from WebSocket ');
+
+        const message = NJSON.parse<WSMessage>(event.data);
+        const { clientId, networkInfo } = message;
+
+        if (clientId) this.logger.info(`Connected with clientId ${(this.clientId = clientId)}`);
+
+        if (networkInfo) {
+          const { eraSummaries, genesisParameters, ledgerTip, lovelaceSupply, protocolParameters, stake } = networkInfo;
+
+          if (eraSummaries) this.networkInfoSubjects.eraSummaries.next(eraSummaries);
+          if (genesisParameters) this.networkInfoSubjects.genesisParameters.next(genesisParameters);
+          if (lovelaceSupply) this.networkInfoSubjects.lovelaceSupply.next(lovelaceSupply);
+          if (protocolParameters) this.networkInfoSubjects.protocolParameters.next(protocolParameters);
+          if (stake) this.networkInfoSubjects.stake.next(stake);
+          // Emit ledgerTip as last one
+          if (ledgerTip) this.networkInfoSubjects.ledgerTip.next(ledgerTip);
+        }
+      } catch (error) {
+        this.logger.error(error, 'While parsing message', event.data, this.clientId);
+      }
+    };
+
+    ws.onclose = () => {
+      this.logger.info('WebSocket client connection closed', this.clientId);
+      this.clientId = NOT_CONNECTED_ID;
+      this.retry();
+    };
+
+    ws.onerror = (error: unknown) => {
+      this.logger.error(isEventError(error) ? error.error : error, 'Async error from WebSocket client', this.clientId);
+      ws.close();
+    };
+
+    ws.onopen = () => {
+      this.status = 'connected';
+      this.ws = ws;
+      this.heartbeat();
+    };
+  }
+
+  private heartbeat() {
+    if (this.heartbeatTimeout) clearInterval(this.heartbeatTimeout);
+
+    this.heartbeatTimeout = setTimeout(() => this.request({}), this.heartbeatInterval);
+    this.heartbeatTimeout.unref();
+  }
+
+  private retry() {
+    if (this.status !== 'stop') {
+      this.status = 'idle';
+      setTimeout(() => this.connect(), 1000).unref();
+    }
+  }
+
+  /** Closes the WebSocket connection. */
+  close() {
+    if (this.heartbeatTimeout) clearInterval(this.heartbeatTimeout);
+
+    this.status = 'stop';
+    this.ws.close();
+  }
+
+  /**
+   * Sends a request through WS to server.
+   *
+   * @param request the request.
+   * @returns `true` is sent, otherwise `false`.
+   */
+  request(request: WSMessage) {
+    if (this.status !== 'connected') return false;
+
+    this.ws!.send(NJSON.stringify(request));
```

Comment:
Can this throw? If yes, then I think it would interrupt the heartbeat interval unrecoverably

Reply:
It can throw only in case or WebSocket disconnection, which has its own handling in the `ws.on('error', ... )` handler, heartbeat included.

Reply:
I'm wondering if an uncaught error here could have any side effects though, it's being called in `setTimeout`. I'd feel safer if we did `try { send(); } catch { return false; }`

Btw, any reason for `request` to be a public method? :thinking:

Reply:
Nice catch! The Promise must be correctly handled in the caller. Thank you.

> any reason for request to be a public method?

Probably just a refuse....

Reply:
6ba4ef5a2d should solve this